distributed/comm/zmq.py

```diff
@@ -0,0 +1,218 @@
+        msg = from_frames([f.buffer for f in frames], deserialize=self.deserialize)
+        raise gen.Return(msg)
+
+    @gen.coroutine
+    def write(self, msg):
+        if self.sock is None:
+            raise CommClosedError
+
+        frames = to_frames(msg)
+        copy = all(len(f) < NOCOPY_THRESHOLD for f in frames)
+        yield self.sock.send_multipart(frames, copy=copy)
+        raise gen.Return(sum(map(len, frames)))
+
+    @gen.coroutine
+    def close(self):
+        sock, self.sock = self.sock, None
+        if sock is not None and not sock.closed:
+            sock.close(linger=5000)   # 5 seconds
+
+    def abort(self):
+        sock, self.sock = self.sock, None
+        if sock is not None and not sock.closed:
+            sock.close(linger=0)      # no wait
+
+    def closed(self):
+        return self.sock is None
+
+
+class ZMQConnector(object):
+
+    @gen.coroutine
+    def _do_connect(self, sock, address, listener_url, deserialize=True):
+        sock.connect(listener_url)
+
+        req = {'op': 'zmq-connect'}
+        yield sock.send_multipart(to_frames(req))
+        frames = yield sock.recv_multipart(copy=True)
+        sock.disconnect(listener_url)
+
+        resp = from_frames(frames)
+        sock.connect(resp['zmq-url'])
+
+        comm = ZMQ(sock, 'zmq://' + address, deserialize)
+        raise gen.Return(comm)
+
+    @gen.coroutine
+    def connect(self, address, deserialize=True):
+        listener_url = make_zmq_url(*parse_host_port(address))
+        sock = make_socket(zmq.DEALER)
+        set_socket_options(sock)
+
+        comm = yield self._do_connect(sock, address, listener_url)
+        raise gen.Return(comm)
+
+
+class ZMQListener(Listener):
+
+    def __init__(self, address, comm_handler, deserialize=True, default_port=0):
+        self.ip, self.port = parse_host_port(address, default_port)
+        self.comm_handler = comm_handler
+        self.deserialize = deserialize
+        self.sock = None
+        self.bound_port = None
+        self.please_stop = False
+
+    def start(self):
+        self.sock = make_socket(zmq.ROUTER)
+        set_socket_options(self.sock)
+        if self.port == 0:
+            self.bound_port = bind_to_random_port(self.sock, self.ip)
+        else:
+            self.bound_port = self.port
+            self.sock.bind(make_zmq_url(self.ip, self.port))
+        self._listen()
+
+    @gen.coroutine
+    def _listen(self):
+        while not self.please_stop:
+            frames = yield self.sock.recv_multipart()
+            envelope = frames[0]
+            req = from_frames(frames[1:])
+            assert req['op'] == 'zmq-connect'
+
+            cli_sock = make_socket(zmq.DEALER)
+            set_socket_options(cli_sock)
+            cli_port = bind_to_random_port(cli_sock, self.ip)
+
+            resp = {'zmq-url': make_zmq_url(self.ip, cli_port)}
+            yield self.sock.send_multipart([envelope] + to_frames(resp))
+
+            address = 'zmq://<unknown>'  # XXX
```

Comment:
zmq calls the routing id 'IDENTITY', and is a uid for routing messages to peers. It can be set manually with `socket.IDENTITY = b'whoiam', but you have to be careful to avoid collisions if you do this. The default is either a monotonic counter or random id.